staging/src/k8s.io/apiserver/pkg/storage/etcd3/linearized_read_test.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/apis/example"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+func TestLinearizedReadRevisionInvariant(t *testing.T) {
+	// The etcd documentation [1] states that "linearized requests must go through the Raft consensus process."
+	// A full round of Raft consensus adds a new item to the Raft log, which is surfaced by etcd as a higher
+	// revision in the response header. Kubernetes exposes this header revision in e.g. List calls, so it is
+	// ultimately client-facing. By default, all the requests that our *etcd3.store{} issues are linearized.
+	// However, this also includes *read* requests, and we would not expect non-mutating requests against etcd
+	// to, by "go[ing] through the Raft consensus process," result in a higher resource version on List calls.
+	// Today, etcd manages to "go through the Raft consensus process" by piggy-backing on other messages, so
+	// linearized reads do *not* bump the key-value store revision. This test exists to ensure that we notice
```

Comment:
I think you're confusing "raft index" and "store revision". Those are distinct things. A linearized read that didn't manage to piggy-back on another message *would* bump the raft index but would *not* bump the store version. Only writes bump the store verison.

In the example in https://github.com/kubernetes/kubernetes/pull/108938#discussion_r844024255, you can see this by doing a list and a dump of endpoint status:

```shell
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json --prefix / | jq .
```

```json
{
  "header": {
    "cluster_id": 16248273405584523000,
    "member_id": 3412179229575531000,
    "revision": 4,
    "raft_term": 2
  },
  "kvs": [
    {
      "key": "L3Rlc3RrZXkx",
      "create_revision": 2,
      "mod_revision": 2,
      "version": 1,
      "value": "dGVzdHZhbHVlMQ=="
    },
    {
      "key": "L3Rlc3RrZXky",
      "create_revision": 3,
      "mod_revision": 3,
      "version": 1,
      "value": "dGVzdHZhbHVlMg=="
    },
    {
      "key": "L3Rlc3RrZXkz",
      "create_revision": 4,
      "mod_revision": 4,
      "version": 1,
      "value": "dGVzdHZhbHVlMw=="
    }
  ],
  "count": 3
}
```

```shell
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370,127.0.0.1:2371,127.0.0.1:2372 etcdctl endpoint status -w json | jq .
```

```json
[
  {
    "Endpoint": "127.0.0.1:2370",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 3412179229575531000,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  },
  {
    "Endpoint": "127.0.0.1:2371",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 1514320713169590800,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  },
  {
    "Endpoint": "127.0.0.1:2372",
    "Status": {
      "header": {
        "cluster_id": 16248273405584523000,
        "member_id": 8920919569373610000,
        "revision": 4,
        "raft_term": 2
      },
      "version": "3.4.18",
      "dbSize": 20480,
      "leader": 3412179229575531000,
      "raftIndex": 12,
      "raftTerm": 2,
      "raftAppliedIndex": 12,
      "dbSizeInUse": 16384
    }
  }
]
```

as you can see, raftIndex (currently at 12) is distinct from store version (currently at 4)

Reply:
I think the test exercises the behavior we want to check, but the comment is implying things that aren't really accurate (that the list resourceVersion depends on the piggy-backing behavior). simplify the comment to "list resourceVersion only changes when writes are performed, not reads"